### DISCUSSION TITLE
emacsclient + magit on FreeBSD 13.0 exec-path issues

### DISCUSSION THREAD

**Discussion by rosso-** [magit/magit#4622] [Support] [answered]
Hello.

I've been using magit for a while, but without emacs as a daemon.
So, when I set up emacs in daemon mode and began to make my edits with emacsclient I got a problem with magit. It couldn't find git binary, because it's absent in exec path (it should include "/usr/local/bin"):
```
exec-path is a variable defined in ‘C source code’.
Its value is
("/sbin" "/bin" "/usr/sbin" "/usr/bin" "/usr/local/libexec/emacs/27.2/amd64-portbld-freebsd13.0" "/usr/local/bin")
Original value was
("/sbin" "/bin" "/usr/sbin" "/usr/bin" "/usr/local/libexec/emacs/27.2/amd64-portbld-freebsd13.0")
```
even more, I can't customize it persistently, just for current session. It's a pity.
```
There is no such problem when I start magit from ordinary emacs process.
```
Maybe magit should import somehow exec-path from emacs at startup?

```
(require 'exec-path-from-shell)
(exec-path-from-shell-initialize)
```
- this doesn't work, as proposed at 
https://emacs.stackexchange.com/questions/27918/why-is-exec-path-different-in-emacsclient-emacsserver-than-in-emacs

**Comment by tarsius** [comment, marked as answer]
I don't think this is something that magit should be concerned with. This affects anything located in /usr/local/bin, not just git as used by magit.

Emacs, like any child process, inherits the environment from its parent process. If you start it in a terminal then it inherits the shells environment, if not, then it doesn't. `exec-path-from-shell` is indeed a popular method for copying variables from the shell configuration to emacs. IMO it would make more sense to go further up the chain and extend PATH in the configuration of the graphical environment, or maybe duplicate the addition, putting them both in the shell config files and the service that starts the emacs demon.

**Reply by rosso-** [reply]
> 
I'm using emacs-nox11 in TUI mode, without Xorg.
Daemon rc startup script (got it from here https://forums.freebsd.org/threads/rc-d-daemon-as-user-emacs.83055/), is:
```
> cat /usr/local/etc/rc.d/emacsd 
#!/bin/sh
#

# PROVIDE: emacsd
# REQUIRE: LOGIN
# KEYWORD: shutdown

. /etc/rc.subr

name="emacsd"
rcvar="${name}_enable"
start_precmd="${name}_prestart"

load_rc_config $name
: "${emacsd_enable:="NO"}"
: "${emacsd_user:="nobody"}"

piddir="/var/run/${emacsd_user}"
pidfile="${piddir}/${name}_daemon.pid"
pidfile_child="${piddir}/${name}_emacs.pid"
emacsd_env="XDG_RUNTIME_DIR=/tmp/$(id -u ${emacsd_user})-runtime-dir HOME=/home/${emacsd_user}"
proc_cmd="/usr/local/bin/emacs"
proc_args="--fg-daemon -u ${emacsd_user}"
command="/usr/sbin/daemon"
command_args="-f -P ${pidfile} -p ${pidfile_child} -r ${proc_cmd} ${proc_args}"

emacsd_prestart()
{
    [ -z "${XDG_RUNTIME_DIR}" ] && XDG_RUNTIME_DIR="/tmp/$(id -u ${emacsd_user})-runtime-dir"
    ! [ -d "${XDG_RUNTIME_DIR}" ] && \
	install -d -o "$(id -u ${emacsd_user})" -m 0700 "${XDG_RUNTIME_DIR}" -

    ! [ -d "${piddir}" ] && \
	install -d -o "$(id -u ${emacsd_user})" -g "$(id -u ${emacsd_user})" "${piddir}" -
    return 0
}

run_rc_command "$1"
```

**Reply by tarsius** [reply]
Please ask elsewhere for help.

I am not familiar with freebsd and its startup scripts but I would suspect you could inject an `export PAHT="..."` in there.
